pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java

```diff
@@ -0,0 +1,381 @@
+import org.broadinstitute.ddp.util.ElasticsearchServiceUtil;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.delete.DeleteRequest;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.client.RestHighLevelClient;
+import org.jdbi.v3.core.Handle;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Deletion of a specified user and data connected to it.
+ * It can be:
+ * <ul>
+ * <li><b>simple deletion</b> - (to be called from a rest service) this kind of deletion deletes not all
+ * data. And before deletion it checks a possibility to delete (for example if a user connected to
+ * auth0 account - if yes, then cannot be deleted).</li>
+ * <li><b>full deletion</b> - (to be called from PubSub) this kind of deletion deletes all user's data
+ * (including the auth0 account).</li>
+ * </ul>
+ *
+ * <p><b>Algorithm:</b>
+ * <ul>
+ *  <li>check if a user not refers to any governed user (if any references exist - cancel deletion);</li>
+ *  <li>delete auth0 data;</li>
+ *  <li>delete kit requests data;</li>
+ *  <li>delete user_study_legacy_data;</li>
+ *  <li>delete user_announcement;</li>
+ *  <li>delete activity_instance_status;</li>
+ *  <li>delete from answer where operator_user_id in (select user_id from delete_users);</li>
+ *  <li>delete user_profile;</li>
+ *  <li>delete user_medical_provider;</li>
+ *  <li>delete user_study_enrollment;</li>
+ *  <li>delete temp_mailing_address, default_mailing_address, mailing_address;</li>
+ *  <li>delete picklist_option__answer, agreement_answer, text_answer, boolean_answer, date_answer, composite_answer_item, answer;</li>
+ *  <li>delete activity_instance_status;</li>
+ *  <li>delete activity_instance;</li>
+ *  <li>delete event_configuration_occurrence_counter;</li>
+ *  <li>delete queued_notification_template_substitution;</li>
+ *  <li>delete queued_notification;</li>
+ *  <li>delete queued_event;</li>
+ *  <li>delete user_governance;</li>
+ *  <li>delete age_up_candidate;</li>
+ *  <li>delete user;</li>
+ *  <li>delete ES data;</li>
+ *  <li>delete auth0;</li>
+ *  <li>send data-sync request.</li>
+ * </ul>
+ */
+public class UserDeleteService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(UserDeleteService.class);
+
+    private static final String REQUEST_TYPE = "_doc";
+
+    private final RestHighLevelClient esClient;
+
+    private static final String ERROR_PREFIX_USER_DELETE = "User [guid=%s] deletion is FAILED: ";
+
+    public UserDeleteService(RestHighLevelClient esClient) {
+        this.esClient = esClient;
+    }
+
+    /**
+     * Used to delete a user which still not registered in auth0-system, not governed by others than a user
+     * who deletes this user, not having status ENROLLED.
+     */
+    public void simpleDelete(Handle handle, User user, String whoDeleted, String comment) throws IOException {
```

Comment:
A full javadoc comment would be useful here, particularly explaining what this arguments mean and do. e.g. whoDeleted: is this just a comment string or an identifier like a guid or an email? I can't tell.

Reply:
Added comments for parameters.